### PR TITLE
Per collection hardware metrics

### PR DESCRIPTION
**Pull request by JojiiOfficial**
Aggregate hardware measurements in the corresponding `Collection` for all measured operations.

Design choices:
  - Utilizes a new wrapper proving by design that no measurements can be dropped without noticing.
  - `HwMeasurementAcc` now also has checks when dropping values in test+debug builds with zero overhead in release, thanks to being completely conditionally compiled.

### CODE REVIEW

File: lib/common/common/src/counter/hardware_accumulator.rs [hunk 1]
```diff
@@ -1,23 +1,42 @@
...
+    #[cfg(any(debug_assertions, test))]
+    applied: Arc<AtomicBool>,
 }
 
 impl HwMeasurementAcc {
     pub fn new() -> Self {
         Self::default()
     }
 
+    pub fn new_unchecked() -> Self {
```
Comment (timvisee) [inline lib/common/common/src/counter/hardware_accumulator.rs:25]: I see that we use this one a lot in this PR. Why can't we use the `new` variant instead for any of them?

File: lib/common/common/src/counter/hardware_accumulator.rs [hunk 2]
```diff
@@ -26,24 +45,125 @@ impl HwMeasurementAcc {
...
+
     /// Consumes and accumulates the values from `hw_counter_cell` into the accumulator.
     pub fn merge_from_cell(&self, hw_counter_cell: impl Into<HardwareCounterCell>) {
         let HardwareCounterCell { ref cpu_counter } = hw_counter_cell.into();
 
         self.cpu_counter
             .fetch_add(cpu_counter.take(), Ordering::Relaxed);
     }
+
+    pub fn has_values(&self) -> bool {
```
Comment (timvisee) [inline lib/common/common/src/counter/hardware_accumulator.rs:81]: Nitpick: how about `is_zero()` (which would be the inverse)?

File: docs/redoc/master/openapi.json [hunk 3]
```diff
@@ -10984,6 +10984,7 @@
         "type": "object",
         "required": [
           "config",
+          "hardware_usage",
```
Comment (timvisee) [inline docs/redoc/master/openapi.json:10987]: Shouldn't this be optional?

File: lib/common/common/src/counter/hardware_accumulator.rs (hunk 1 above)
Comment (JojiiOfficial) [reply lib/common/common/src/counter/hardware_accumulator.rs:25]: In the Drop-function of this type we have a check that helps us not to accidentally lose measurements in debug mode and tests.
This function makes it clear that when creating an instance of `HwMeasurementAcc`, we don't want it to have this check, which makes writing tests or hardware unrelated code easier to work with.

File: docs/redoc/master/openapi.json (hunk 3 above)
Comment (JojiiOfficial) [reply docs/redoc/master/openapi.json:10987]: This is the telemetry endpoint and as far as I understood, we wanted it to be able to turn off for each API that gets measured (like search or matrix), but not telemetry.
I could be wrong here though.

Maybe @generall can give his opinion, whether we want it always to be in telemetry endpoint, or only in a certain condition.

Comment (generall) [reply docs/redoc/master/openapi.json:10987]: It should be option at least for backward compatibility of the clients. If new client reads old telemetry output it should not fail

File: lib/collection/src/collection/collection_ops.rs [hunk 4]
```diff
@@ -341,7 +341,11 @@ impl Collection {
                     .copied()
                     .unwrap_or(ReplicaState::Dead);
                 let count_result = replica_set
-                    .count_local(count_request.clone(), None, HwMeasurementAcc::new())
+                    .count_local(
+                        count_request.clone(),
+                        None,
+                        HwMeasurementAcc::new_unchecked(),
```
Comment (generall) [inline lib/collection/src/collection/collection_ops.rs:347]: This needs a comment why `unchecked` is acceptable.

Comment (generall) [reply lib/collection/src/collection/collection_ops.rs:347]: Since this is not a test and not a bench

File: lib/collection/src/collection/mod.rs [hunk 5]
```diff
@@ -176,6 +178,7 @@ impl Collection {
             update_runtime: update_runtime.unwrap_or_else(Handle::current),
             search_runtime: search_runtime.unwrap_or_else(Handle::current),
             optimizer_cpu_budget,
+            hardware_usage: HwMeasurementAcc::new_unchecked(),
```
Comment (generall) [inline lib/collection/src/collection/mod.rs:181]: Why `unchecked` is acceptable?

File: lib/collection/src/collection/mod.rs [hunk 6]
```diff
@@ -285,6 +288,7 @@ impl Collection {
...
```
Comment (generall) [inline lib/collection/src/collection/mod.rs:291]: Why `unchecked` is acceptable?

File: lib/collection/src/shards/transfer/resharding_stream_records.rs [hunk 7]
```diff
@@ -70,7 +70,7 @@ pub(crate) async fn transfer_resharding_stream_records(
                     exact: true,
                 }),
                 None,
-                HwMeasurementAcc::new(),
+                HwMeasurementAcc::new_unchecked(),
```
Comment (generall) [inline lib/collection/src/shards/transfer/resharding_stream_records.rs:73]: needs a comment

**Comment by JojiiOfficial** [comment]
Closing in favor of #5453
